node/core/av-store/src/lib.rs

```diff
@@ -110,62 +348,201 @@ impl AvailabilityStoreSubsystem {
+						}
+					}
+					Ok(FromOverseer::Signal(OverseerSignal::BlockFinalized(hash))) => {
+						process_block_finalized(&subsystem, ctx, &subsystem.inner, hash).await?;
+					}
 					Ok(FromOverseer::Communication { msg }) => {
-						process_message(&subsystem.inner, &subsystem.metrics, msg)?;
+						process_message(&mut subsystem, ctx, msg).await?;
 					}
 					Err(_) => break,
 				}
 			}
+			pov_pruning_time = pov_pruning_time => {
+				subsystem.prune_povs()?;
+			}
+			chunk_pruning_time = chunk_pruning_time => {
+				subsystem.prune_chunks()?;
+			}
 			complete => break,
 		}
 	}
 
 	Ok(())
 }
 
-fn process_message(db: &Arc<dyn KeyValueDB>, metrics: &Metrics, msg: AvailabilityStoreMessage) -> Result<(), Error> {
+/// As soon as certain block is finalized its pruning records and records of all
+/// blocks that we keep that are `older` than the block in question have to be updated.
+///
+/// The state of data has to be changed from
+/// `CandidateState::Included` to `CandidateState::Finalized` and their pruning times have
+/// to be updated to `now` + keep_finalized_{block, chunk}_for`.
+async fn process_block_finalized<Context>(
+	subsystem: &AvailabilityStoreSubsystem,
+	ctx: &mut Context,
+	db: &Arc<dyn KeyValueDB>,
+	hash: Hash,
+) -> Result<(), Error>
+where
+	Context: SubsystemContext<Message=AvailabilityStoreMessage>
+{
+	let block_number = get_block_number(ctx, hash).await?;
+
+	if let Some(mut pov_pruning) = pov_pruning(db) {
+		// Since the records are sorted by time in which they need to be pruned and not by block
+		// numbers we have to iterate through the whole collection here.
+		for record in pov_pruning.iter_mut() {
+			if record.block_number <= block_number {
+				log::trace!(
+					target: LOG_TARGET,
+					"Updating pruning record for finalized block {}",
+					record.candidate_hash,
+				);
+
+				record.prune_at = SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs() +
+					subsystem.pruning_config.keep_finalized_block_for
+			}
+		}
+
+		// has to be kept sorted.
+		pov_pruning.sort();
+
+		put_pov_pruning(db, None, pov_pruning)?;
+	}
+
+	Ok(())
+}
+
+async fn process_block_activated<Context>(
+	ctx: &mut Context,
+	db: &Arc<dyn KeyValueDB>,
+	hash: Hash,
+) -> Result<(), Error>
+where
+	Context: SubsystemContext<Message=AvailabilityStoreMessage>
+{
+	let events = request_candidate_events(ctx, hash).await?;
+
+	log::trace!(target: LOG_TARGET, "block activated {}", hash);
+	if let Some(mut pov_pruning) = pov_pruning(db) {
+		let mut included = HashSet::new();
+
+		for event in events.into_iter() {
+			if let CandidateEvent::CandidateIncluded(receipt, _) = event {
+				log::trace!(target: LOG_TARGET, "Candidate {} was included", receipt.hash());
+				included.insert(receipt.hash());
+			}
+		}
+
+		for record in pov_pruning.iter_mut() {
+			if included.contains(&record.candidate_hash) {
+				record.prune_at = std::u64::MAX;
+			}
+		}
+
+		pov_pruning.sort();
+
+		put_pov_pruning(db, None, pov_pruning)?;
```

Comment:
Given sorting on a presorted array is `O(n)` ( https://doc.rust-lang.org/src/alloc/slice.rs.html#192-197 ) we should consider always sorting when calling `put_` in order to guarantee the invariant (`fn prune_povs` would technically not require this)